onnxruntime/contrib_ops/cpu/quantization/moe_quantization_cpu.cc

```diff
@@ -0,0 +1,438 @@
+    constexpr float swiglu_alpha = 1.702f;
+    for (int64_t i = 0; i < inter_size; ++i) {
+      float linear_val = fc1_output[i];             // First half: linear projection
+      float gate_val = fc1_output[i + inter_size];  // Second half: gate projection
+      // SwiGLU: gate * sigmoid(alpha * gate) * (linear + 1)
+      float sigmoid_arg = swiglu_alpha * gate_val;
+      float sigmoid_out = 1.0f / (1.0f + std::exp(-sigmoid_arg));
+      float swish_out = gate_val * sigmoid_out;
+      result[i] = swish_out * (linear_val + 1.0f);
+    }
+  };
+
+  // Determine quantization parameters based on bit width
+  const bool is_4bit = UseUInt4x2;
+  const float zero_point = is_4bit ? 8.0f : 128.0f;
+  const int64_t act_multiplier = is_swiglu ? 2 : 1;
+  const int64_t fc1_output_size = is_swiglu ? 2 * moe_params.inter_size : moe_params.inter_size;
+
+  // Calculate weight sizes and strides based on quantization type
+  const int64_t fc1_weight_stride = is_4bit ? (moe_params.hidden_size * fc1_output_size / 2) : (moe_params.hidden_size * moe_params.inter_size * act_multiplier);
+  const int64_t fc2_weight_stride = is_4bit ? (moe_params.inter_size * moe_params.hidden_size / 2) : (moe_params.inter_size * moe_params.hidden_size);
+
+  // Pre-dequantize all expert weights once (shared across all threads)
+  auto dequant_fc1_weights = IAllocator::MakeUniquePtr<float>(allocator,
+                                                              static_cast<size_t>(moe_params.num_experts * moe_params.hidden_size * (is_4bit ? fc1_output_size : moe_params.inter_size * act_multiplier)));
+  auto dequant_fc2_weights = IAllocator::MakeUniquePtr<float>(allocator,
+                                                              static_cast<size_t>(moe_params.num_experts * moe_params.inter_size * moe_params.hidden_size));
+
+  // Helper lambda for dequantizing a single weight value
+  auto DequantizeWeight = [&](const uint8_t* weights, size_t weight_idx, size_t linear_idx,
+                              const float* scales, int64_t scale_idx) -> float {
+    if (is_4bit) {
+      // For Int4, two values are packed in each uint8
+      size_t packed_idx = linear_idx / 2;
+      uint8_t packed_value = weights[packed_idx];
+      uint8_t quantized_weight = (linear_idx % 2 == 0) ? (packed_value & 0x0F) : ((packed_value >> 4) & 0x0F);
+      return (static_cast<float>(quantized_weight) - zero_point) * scales[scale_idx];
+    } else {
+      // For Int8, direct access
+      return (static_cast<float>(weights[weight_idx]) - zero_point) * scales[scale_idx];
+    }
+  };
+
+  // Dequantize FC1 weights for all experts
+  concurrency::ThreadPool::TryParallelFor(
+      thread_pool, static_cast<std::ptrdiff_t>(moe_params.num_experts),
+      static_cast<double>(std::max<int64_t>(1, moe_params.num_experts / num_threads)),
+      [&](ptrdiff_t expert_start, ptrdiff_t expert_end) {
+        for (int64_t expert_idx = expert_start; expert_idx < expert_end; ++expert_idx) {
```

Comment:
Keeping the type as `std::ptrdiff_t` is safer.

```suggestion
        for (std::ptrdiff_t expert_idx = expert_start; expert_idx < expert_end; ++expert_idx) {
```

We can then use `SafeInt` to prevent integer overflow when using `expert_idx`.